parlai/core/worlds.py

```diff
@@ -1499,6 +1533,622 @@ def shutdown(self):
+                # Update
+                self.buffers[worker_idx][key] = buff
+            for key, new_size in batch_obs.resized.items():
+                self.buffers[worker_idx][key].resize_(new_size)
+            for key, buff in self.buffers[worker_idx].items():
+                setattr(batch_obs, key, buff)
+        return batch_obs
+
+    def parley(self):
+        """
+        QueueWorld parley.
+
+        QueueWorld receives a batch from its produce queue, gives it to an agent to act,
+        then puts it back on the appropriate consume queue.
+        """
+        self._maybe_begin_processes()
+        self._maybe_parley_init()
+        queue_result, batch_obs, worker_idx = self._poll_produce_queue()
+        if self.epoch_done():
+            return
+        batch_obs = self._maybe_handle_buffers(batch_obs, worker_idx)
+        batch_acts = self.batch_act(batch_obs)
+        self.consume_queues[worker_idx].put_nowait((QueueSignal.BATCH, batch_acts))
+        self.update_counters(batch_obs)
+
+    def batch_act(
+        self, batch_observation: Union[List[Message], Batch]
+    ) -> List[Message]:
+        """
+        Batch act on agent.
+
+        :param batch_observation:
+            The observation that the agent will act upon
+
+        :return: batch_actions
+        :rtype: List[Message]
+        """
+        a = self.world.get_model_agent()
+        if hasattr(a, 'batch_act'):
+            batch_actions = a.batch_act(batch_observation)
+        else:
+            batch_actions = []
+            for w in self.worlds:
+                agents = w.get_agents()
+                acts = w.get_acts()
+                acts[1] = agents[1].act()
+                batch_actions.append(acts[1])
+        return batch_actions
+
+    def shutdown(self):
+        """
+        Shutdown subworlds and close all queues.
+        """
+        for w in self.worlds:
+            w.shutdown()
+        self.world.shutdown()
+        for q in self.consume_queues:
+            q.put_nowait((QueueSignal.TERMINATE, None))
+        for q in [self.produce_queue, self.report_queue] + self.consume_queues:
+            q.close()
+        for p in self.processes:
+            p.terminate()
+
+    def reset(self):
+        """
+        Reset each subworld and additionally each individual process.
+        """
+        super().reset()
+        self.world.reset()
+        for w in self.worlds:
+            w.reset()
+        while True:
+            try:
+                self.produce_queue.get_nowait()
+            except queue.Empty:
+                break
+        for q in self.consume_queues:
+            q.put((QueueSignal.RESET, None))
+        self.finished_workers = []
+        self.init_parley = True
+
+    def reset_metrics(self):
+        """
+        Reset metrics.
+
+        (TODO possibly reset process metrics)
+        """
+        super().reset_metrics()
+        self.world.reset_metrics()
+        # for q in self.consume_queues:
```

Comment:
Heads up

Reply:
i actually need to uncomment this thanks